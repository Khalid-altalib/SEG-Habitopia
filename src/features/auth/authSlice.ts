import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import { LocalUser } from "../../../types";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { RootState } from "../../app/store";
import { Toast } from "react-native-toast-message/lib/src/Toast";
import { createUserInDatabase } from "./authQueries";
import { Auth } from "aws-amplify";

type AuthState = {
  signUpData: {
    email: string;
    password: string;
    name: string;
    confirmationCode: string;
  };
  logInData: {
    email: string;
    password: string;
  };
  user: LocalUser | undefined;
  loading: boolean;
  error: string;
};

export const signUpUser = createAsyncThunk<void, void, { rejectValue: string }>(
  "auth/signUp",
  async (_, thunkAPI) => {
    try {
      const state = thunkAPI.getState() as RootState;
      const { email, password, name } = state.auth.signUpData;
      await Auth.signUp({
        username: email as string,
        password: password as string,
        attributes: {
          name: name,
        },
      });
    } catch (error: any) {
      const message = error.message;
      console.log(error);
      return thunkAPI.rejectWithValue(message);
    }
  }
);

export const sendConfirmationCode = createAsyncThunk<
  LocalUser,
  void,
  { rejectValue: string }
>("auth/confirmation", async (_, thunkAPI) => {
  try {
    const state = thunkAPI.getState() as RootState;
    const { email, password, confirmationCode, name } = state.auth.signUpData;
    await Auth.confirmSignUp(email, confirmationCode);
    const user = await logInHelper(email, password, name);
    return user as LocalUser;
  } catch (error: any) {
    const message = error.message;
    console.log(error);
    return thunkAPI.rejectWithValue(message);
  }
});

export const updatePassword = async (password: string, thunkAPI: any) => {
  console.log("updating password");
  console.log(password);
  const oldPassword = "Password.123";
  const state = thunkAPI.getState() as RootState;
  // console.log(state);
  const email = state.settings.settings.email;
  console.log(email);

  Auth.currentAuthenticatedUser()
  .then((user) => {
    return Auth.changePassword(user, oldPassword, password);
  })
  .then((data) => console.log(data))
  .catch((err) => console.log(err));
  
};

const logInHelper = async (email: string, password: string, name?: string) => {
  const { signInUserSession, attributes } = await Auth.signIn(email, password);
  const user = {
    authToken: signInUserSession.idToken.jwtToken,
    userId: attributes.sub,
  };

  await AsyncStorage.setItem("user", JSON.stringify(user));
  await createUserInDatabase(attributes.sub, email, name);
  return user;
};

export const logInUser = createAsyncThunk<
  LocalUser,
  void,
  { rejectValue: string }
>("auth/login", async (_, thunkAPI) => {
  try {
    const state = thunkAPI.getState() as RootState;
    const { email, password } = state.auth.logInData;
    const user = await logInHelper(email, password);
    return user as LocalUser;
  } catch (error: any) {
    const message = error.message;
    console.log(message);
    return thunkAPI.rejectWithValue(message);
  }
});

export const logInUserFromStorage = createAsyncThunk<
  LocalUser,
  void,
  { rejectValue: string }
>("auth/logInFromStorage", async (_, thunkAPI) => {
  const data = await AsyncStorage.getItem("user");

  if (data) {
    Toast.show({
      type: "success",
      text1: "Welcome back! ðŸ‘‹",
      text2: "We've logged you in from your previous session",
    });
    return JSON.parse(data) as LocalUser;
  } else {
    return thunkAPI.rejectWithValue("Please log in again");
  }
});

export const logOutUser = createAsyncThunk("auth/logOutUser", async () => {
  await AsyncStorage.removeItem("user");
  return undefined;
});

const initialState: AuthState = {
  signUpData: { email: "", password: "", name: "", confirmationCode: "" },
  logInData: { email: "", password: "" },
  user: undefined,
  loading: false,
  error: "",
};

const authenticationPending = (state: AuthState) => {
  state.loading = true;
};

const authenticationFulfilled = (
  state: AuthState,
  action: PayloadAction<LocalUser>
) => {
  state.loading = false;
  state.user = action.payload;
  state.error = "";
};

const authenticationRejected = (state: AuthState, action: any) => {
  state.loading = false;
  state.user = undefined;
  state.error = action.payload;
};

export const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    addSignUpData: (state, action: PayloadAction<object>) => {
      console.log(state, action);
      state.signUpData = { ...state.signUpData, ...action.payload };
    },
    addLogInData: (state, action: PayloadAction<object>) => {
      state.logInData = { ...state.logInData, ...action.payload };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(logInUser.pending, authenticationPending);
    builder.addCase(logInUser.fulfilled, authenticationFulfilled);
    builder.addCase(logInUser.rejected, authenticationRejected);
    builder.addCase(sendConfirmationCode.pending, authenticationPending);
    builder.addCase(sendConfirmationCode.fulfilled, authenticationFulfilled);
    builder.addCase(sendConfirmationCode.rejected, authenticationRejected);
    builder.addCase(logInUserFromStorage.pending, authenticationPending);
    builder.addCase(logInUserFromStorage.fulfilled, authenticationFulfilled);
    builder.addCase(logInUserFromStorage.rejected, authenticationRejected);
    builder.addCase(logOutUser.pending, (state) => {
      state.loading = true;
    });
    builder.addCase(logOutUser.fulfilled, (state) => {
      state.loading = false;
      state.user = undefined;
    });
  },
});

export const { addSignUpData, addLogInData } = authSlice.actions;

export default authSlice.reducer;
