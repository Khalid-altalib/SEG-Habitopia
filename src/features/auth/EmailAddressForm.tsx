// Native Base
import { Input } from "native-base";

// React Native
import { View } from "react-native";

// React Navigation
import { NavigationProp, useNavigation } from "@react-navigation/native";

// React Hook Form
import { useForm, Controller } from "react-hook-form";

// Expo
import { AntDesign } from "@expo/vector-icons";

// Habitopia
import { addSignUpData } from "@features/auth/authSlice";
import { useDispatch } from "@app/hooks";
import Button from "@components/Button";
import Text from "@components/Text";
import { AuthParams, ButtonType, TextType } from "types";

type formData = {
  email: string;
};

/**
 * A form which allows the user to enter their email address,
 * adds that data to the Redux store, and then navigates to
 * the password screen.
 *
 * @returns - The email address form component.
 */
const EmailAddressForm = () => {
  // The initial state of the form data.
  const { control, handleSubmit } = useForm({
    defaultValues: {
      email: "",
    },
  });

  // A handler for when the user submits the form.
  const dispatch = useDispatch();

  const onSubmit = async (data: formData) => {
    dispatch(addSignUpData(data));
    navigation.navigate("Password");
  };

  const navigation = useNavigation<NavigationProp<AuthParams>>();

  return (
    <View>
      <Text style={{ marginBottom: 5 }} type={TextType.Subtle}>
        Email Address
      </Text>
      <Controller
        control={control}
        render={({ field: { onChange, value } }) => (
          <Input onChangeText={onChange} value={value} size="xl" />
        )}
        name="email"
      />
      <Button
        onPress={handleSubmit(onSubmit)}
        type={ButtonType.Primary}
        style={{ marginTop: 20 }}
        isFullWidth
        icon={<AntDesign name="arrowright" size={20} color="white" />}
      >
        Continue
      </Button>
    </View>
  );
};

export default EmailAddressForm;
