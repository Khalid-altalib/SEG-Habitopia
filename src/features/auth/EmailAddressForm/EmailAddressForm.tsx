import React from "react";
import { Checkbox, Input } from "native-base";
import { useForm, Controller } from "react-hook-form";
import { AntDesign } from "@expo/vector-icons";

import { addLogInData, addSignUpData, logInUser } from "../authSlice";
import { useDispatch } from "../../../app/hooks";
import Button from "../../../components/Button";
import Text from "../../../components/Text";
import { AuthParams, ButtonType, TextType } from "../../../../types";
import { View } from "react-native";
import { NavigationProp, useNavigation } from "@react-navigation/native";
import { Toast } from "react-native-toast-message/lib/src/Toast";

type formData = {
  email: string;
};

const EmailAddressForm = () => {
  const { control, handleSubmit } = useForm({
    defaultValues: {
      email: "",
    },
  });
  const dispatch = useDispatch();

  const onSubmit = async (data: formData) => {
    if (!data.email.includes("@")) {
      Toast.show({
        type: "error",
        text1: "Please ensure email is valid",
      });
    } else {
      dispatch(addSignUpData(data));
      navigation.navigate("Password");
    }
  };

  const navigation = useNavigation<NavigationProp<AuthParams>>();

  return (
    <View>
      <Text style={{ marginBottom: 5 }} type={TextType.Subtle}>
        Email Address
      </Text>
      <Controller
        control={control}
        render={({ field: { onChange, value } }) => (
          <Input
            onChangeText={onChange}
            value={value}
            size="xl"
            testID="email-input"
          />
        )}
        name="email"
      />
      <Button
        onPress={handleSubmit(onSubmit)}
        type={ButtonType.Primary}
        style={{ marginTop: 20 }}
        isFullWidth
        icon={<AntDesign name="arrowright" size={20} color="white" />}
      >
        Continue
      </Button>
    </View>
  );
};

export default EmailAddressForm;
