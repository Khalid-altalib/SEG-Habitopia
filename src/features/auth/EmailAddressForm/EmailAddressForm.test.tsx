import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import configureStore from "redux-mock-store";
import EmailAddressForm from "./EmailAddressForm";
import TestingWrapperNavigation from "@app/testingWrapperWithNavigation";
import thunk from "redux-thunk";
import { addSignUpData } from "../authSlice";

const mockStore = configureStore([thunk]);
const store = mockStore({});

describe("EmailAddressForm", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = render(
      <TestingWrapperNavigation store={store}>
        <EmailAddressForm />
      </TestingWrapperNavigation>
    );
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("dispatches addSignUpData when the user submits the form", async () => {
    // Simulate user input
    const emailInput = wrapper.getByTestId("email-input");
    fireEvent.changeText(emailInput, "john@example.org");

    // Simulate submit button press
    const submitButton = wrapper.getByText("Continue");
    await fireEvent.press(submitButton);

    // Wait for any asynchronous actions to finish
    await waitFor(() => {
      expect(store.getActions()).toContainEqual(
        addSignUpData({ email: "john@example.org" })
      );
    });
  });
});
