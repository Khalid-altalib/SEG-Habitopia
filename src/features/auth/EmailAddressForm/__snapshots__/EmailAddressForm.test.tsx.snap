// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailAddressForm renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View>
              <Text
                style={
                  [
                    {
                      "color": "#18181b",
                    },
                    {
                      "color": "#71717a",
                      "fontFamily": undefined,
                      "fontSize": 15,
                    },
                    {
                      "marginBottom": 5,
                    },
                  ]
                }
              >
                Email Address
              </Text>
              <View
                dataSet={{}}
                isFocused={false}
                style={
                  {
                    "alignItems": "center",
                    "borderColor": "#d4d4d4",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "flexDirection": "row",
                    "overflow": "hidden",
                  }
                }
              >
                <TextInput
                  accessible={true}
                  dataSet={{}}
                  editable={true}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  onKeyPress={[Function]}
                  placeholderTextColor="#a3a3a3"
                  secureTextEntry={false}
                  style={
                    {
                      "backgroundColor": "transparent",
                      "color": "#171717",
                      "flex": 1,
                      "fontFamily": undefined,
                      "fontSize": 18,
                      "fontStyle": "normal",
                      "fontWeight": "400",
                      "height": "100%",
                      "paddingBottom": 8,
                      "paddingLeft": 12,
                      "paddingRight": 12,
                      "paddingTop": 8,
                      "width": "100%",
                    }
                  }
                  testID="email-input"
                  value=""
                />
              </View>
              <View
                style={
                  [
                    {
                      "borderRadius": 10,
                      "shadowColor": "black",
                      "shadowOffset": {
                        "height": 5,
                        "width": 5,
                      },
                      "shadowOpacity": 0.15,
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "marginTop": 20,
                      "opacity": 1,
                    }
                  }
                >
                  <ViewManagerAdapter_ExpoLinearGradient
                    proxiedProperties={
                      {
                        "colors": [
                          4278220774,
                          4286456526,
                        ],
                        "endPoint": [
                          1,
                          1,
                        ],
                        "locations": undefined,
                        "startPoint": [
                          0,
                          0,
                        ],
                      }
                    }
                    style={
                      [
                        {
                          "alignItems": "center",
                          "borderRadius": 10,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "paddingHorizontal": 40,
                          "paddingVertical": 15,
                        },
                        {
                          "width": "100%",
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#18181b",
                          },
                          {
                            "color": "#e4e4e7",
                            "fontFamily": undefined,
                            "fontSize": 19,
                          },
                          [
                            null,
                            {
                              "alignSelf": "center",
                            },
                          ],
                        ]
                      }
                    >
                      Continue
                    </Text>
                    <View
                      style={
                        {
                          "marginLeft": 5,
                        }
                      }
                    >
                      <Text />
                    </View>
                  </ViewManagerAdapter_ExpoLinearGradient>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;
